# Exercise 4.9 - Part 1

Import your analysis libraries, as well as your new customer data set as a dataframe.

## Question 4:
Data wrangling:
- checked the data types, changed some of the column names and dropped the fam_status column. 

## Question 5:
Checked for:
- data types
- missing values
- duplicates

## Question 6:
- combine the customer_data and the ords_prods_merge on their user_id



## Question 3:
- Import your analysis libraries, as well as your new customer data set as a dataframe.

In [47]:
# import libraries
import pandas as pd
import numpy as np
import os
import matplotlib.pyplot as plt
import seaborn as sns
import scipy

In [48]:
# import new customer data
path = r'/Users/azadehrezaei/Desktop/Career Foundry/Achievement 4/08-2021 Instacart Basket Analysis'
customer_data = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'customers.csv'), index_col = False)

## Question 4:
- Wrangle the data so that it follows consistent logic; for example, rename columns with illogical names and drop columns that don’t add anything to your analysis.

I originally wanted to remove the name (first and last) columns due to privacy concerns; but then I decided to keep it in case at some point we decided to reward a specific customer becuase of their purchases and we needed their info. I decided to keep all columns except for the fam_status. I don't think it adds any value to our analysis or makes any difference; We can already get all of our needed information needed for profiling our customers from age, gender, n_dependents and income, so the fam_status doesn't seem necessary. 
In the wrangling process, I also decided to rename some of the columns for more clarity and also for creating uniform titles. 


In [49]:
customer_data.head()

user_id First Name    Surnam  Gender       STATE  Age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   n_dependants fam_status  income  
0             3    married  165665  
1             0     single   59285  
2             2    married   99568  
3             0     single   42049  
4             1    married   40374

In [50]:
customer_data.shape

(206209, 10)

In [51]:
customer_data.columns

Index(['user_id', 'First Name', 'Surnam', 'Gender', 'STATE', 'Age',
       'date_joined', 'n_dependants', 'fam_status', 'income'],
      dtype='object')

In [52]:
customer_data.describe()

user_id            Age   n_dependants         income
count  206209.000000  206209.000000  206209.000000  206209.000000
mean   103105.000000      49.501646       1.499823   94632.852548
std     59527.555167      18.480962       1.118433   42473.786988
min         1.000000      18.000000       0.000000   25903.000000
25%     51553.000000      33.000000       0.000000   59874.000000
50%    103105.000000      49.000000       1.000000   93547.000000
75%    154657.000000      66.000000       3.000000  124244.000000
max    206209.000000      81.000000       3.000000  593901.000000

- Data Wrangling: change the name of some of the columns (First Name, Surnam, Gender, STATE, Age)

In [53]:
# Changing the name of the columns:
customer_data = customer_data.rename(columns = {'First Name':'first_name', 'Surnam':'last_name', 'Gender':'gender', 'STATE':'state', 'Age':'age'})

In [54]:
customer_data.head()

user_id first_name last_name  gender       state  age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   n_dependants fam_status  income  
0             3    married  165665  
1             0     single   59285  
2             2    married   99568  
3             0     single   42049  
4             1    married   40374

## Question 5: 
- Complete the fundamental data quality and consistency checks you’ve learned throughout this Achievement; for example, check for and address missing values and duplicates, and convert any mixed-type data.

I checked the data type for the columns in the dataframe, I decided to change the data type of the user_id to object, because althoughthe user_id is a number, it's an identifier and we don't treat it as an integer. 

In [55]:
# checkign data types:
customer_data.dtypes

user_id          int64
first_name      object
last_name       object
gender          object
state           object
age              int64
date_joined     object
n_dependants     int64
fam_status      object
income           int64
dtype: object

In [56]:
customer_data['user_id'] = customer_data['user_id'].astype('str')

In [57]:
customer_data.dtypes

user_id         object
first_name      object
last_name       object
gender          object
state           object
age              int64
date_joined     object
n_dependants     int64
fam_status      object
income           int64
dtype: object

In [58]:
# checking for missing values:
customer_data.isnull().sum()

user_id             0
first_name      11259
last_name           0
gender              0
state               0
age                 0
date_joined         0
n_dependants        0
fam_status          0
income              0
dtype: int64

- Based on the results of the above query, the first name column has 11,259 missing values. This is interesting because there are no missing values in the last_name column. These missing values make up 5.45% of the total number of the rows. For this reason I decided to replace these missing values with "no_name". 

In [59]:
# fill in the missing values in the first_name column with no_name:
customer_data['first_name'].fillna("no_name", inplace = True) 

In [60]:
# checking for the missing values:
customer_data.isnull().sum()

user_id         0
first_name      0
last_name       0
gender          0
state           0
age             0
date_joined     0
n_dependants    0
fam_status      0
income          0
dtype: int64

- Checking for Duplicates:
I checked for duplicates and did not find any! Yay!

In [61]:
# checking for duplicates:
customer_dups = customer_data[customer_data.duplicated()]

In [62]:
customer_dups.shape

(0, 10)

## Question 6:
- Combine your customer data with the rest of your prepared Instacart data. Tip: Make sure the key columns are of the same data type!

In [63]:
ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', '4.8_ords_prods_merge.pkl'))

In [64]:
ords_prods_merge.head()

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  user_id  order_number  order_day_of_week  order_hour_of_day  ...  \
0   3139998      138            28                  6                 11  ...   
1   1977647      138            30                  6                 17  ...   
2    389851      709             2                  0                 21  ...   
3    652770      764             1                  3                 13  ...   
4   1813452      764             3                  4                 17  ...   

     price_range_loc     busiest_day  slowest/busiest days  \
0  Mid-range Product  Regularly busy          Regular Busy   
1  Mid-range Product  Regularly busy          Regular Busy   
2  Mid-range Product     Busiest day          Buesiest Day   
3  Mid-range Product  Regularly busy            Least Busy   
4  Mid-range Product      Least busy            Least Busy   

  busiest_period_of_day max_order      loyalty_flag avg_price spending_flag  \
0           Most Orders        32  Regular customer  6.935811   Low Spender   
1        Average Orders        32  Regular customer  6.935811   Low Spender   
2        Average Orders         5      New customer  7.930208   Low Spender   
3           Most Orders         3      New customer  4.972414   Low Spender   
4        Average Orders         3      New customer  4.972414   Low Spender   

   order_freq     frequency_flag  
0         8.0  Frequent Customer  
1         8.0  Frequent Customer  
2         6.0  Frequent Customer  
3         9.0  Frequent Customer  
4         9.0  Frequent Customer  

[5 rows x 24 columns]

In [41]:
ords_prods_merge.dtypes

product_id                   int64
product_name                object
aisle_id                     int64
department_id                int64
prices                     float64
order_id                     int64
user_id                      int64
order_number                 int64
order_day_of_week            int64
order_hour_of_day            int64
days_since_prior_order     float64
add_to_cart_order            int64
reordered                    int64
_merge                    category
price_range_loc             object
busiest_day                 object
slowest/busiest days        object
busiest_period_of_day       object
max_order                    int64
loyalty_flag                object
avg_price                  float64
spending_flag               object
order_freq                 float64
frequency_flag              object
dtype: object

In [65]:
ords_prods_merge['user_id'] = ords_prods_merge['user_id'].astype('str')

In [66]:
ords_prods_merge.dtypes

product_id                   int64
product_name                object
aisle_id                     int64
department_id                int64
prices                     float64
order_id                     int64
user_id                     object
order_number                 int64
order_day_of_week            int64
order_hour_of_day            int64
days_since_prior_order     float64
add_to_cart_order            int64
reordered                    int64
_merge                    category
price_range_loc             object
busiest_day                 object
slowest/busiest days        object
busiest_period_of_day       object
max_order                    int64
loyalty_flag                object
avg_price                  float64
spending_flag               object
order_freq                 float64
frequency_flag              object
dtype: object

In [67]:
customer_data.head()

user_id first_name last_name  gender       state  age date_joined  \
0   26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1   33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2   65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3  125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4  130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   n_dependants fam_status  income  
0             3    married  165665  
1             0     single   59285  
2             2    married   99568  
3             0     single   42049  
4             1    married   40374

- I am merging the two dataframes, customer_data & ords_prods_merge on their user_id.

In [68]:
# merge the two dataframes:
df_merged = ords_prods_merge.merge(customer_data, on =['user_id'])

In [69]:
df_merged.head()

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2         907        Premium Sliced Bacon       106             12    20.0   
3         907        Premium Sliced Bacon       106             12    20.0   
4        1000                    Apricots        18             10    12.9   

   order_id user_id  order_number  order_day_of_week  order_hour_of_day  ...  \
0   3139998     138            28                  6                 11  ...   
1   1977647     138            30                  6                 17  ...   
2   3160996     138             1                  5                 13  ...   
3   2254091     138            10                  5                 14  ...   
4    505689     138             9                  6                 12  ...   

      frequency_flag  first_name  last_name gender      state age date_joined  \
0  Frequent Customer     Charles        Cox   Male  Minnesota  81    8/1/2019   
1  Frequent Customer     Charles        Cox   Male  Minnesota  81    8/1/2019   
2  Frequent Customer     Charles        Cox   Male  Minnesota  81    8/1/2019   
3  Frequent Customer     Charles        Cox   Male  Minnesota  81    8/1/2019   
4  Frequent Customer     Charles        Cox   Male  Minnesota  81    8/1/2019   

  n_dependants  fam_status income  
0            1     married  49620  
1            1     married  49620  
2            1     married  49620  
3            1     married  49620  
4            1     married  49620  

[5 rows x 33 columns]

## Question 8:

- Export the new dataframe as a pickle file for the second part of this exercise:

In [70]:
# export data:
df_merged.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', '4.9_ords_prods_customers.pkl'))